crates/scope/src/builder.rs

```diff
@@ -1887,11 +1887,24 @@ impl ScopeKindStack {
 #[derive(Debug)]
 struct ScopeBuilderStack {
     stack: Vec<ScopeBuilder>,
+
+    /// List of names defined and closed over by inner script.
+    /// List is delimited by `None`, for each scope.
+    ///
+    /// The order of scopes are depth-first post-order, and names inside each
+    /// scope is in not defined.
+    ///
+    /// The content of this vector is consumed when leaving each function,
+    /// and reset to empty.
```

Comment:
I do not understand how this could work properly.

If I understand correctly, one scope of an outer function might considered as part of the closed over bindings of the inner function when we `add_closed_over_bindings`, unless the fact of doing a reset after leaving a function is completely arbitrary?

Reply:
we propagate free variables in `FreeNameTracker`, and when leaving function, all free vars are converted to possibly-closed-over names.

and when we leave any scope, names that 1. defined in the scope, and 2. closed over by inner scope, are added to this list.

Reply:
My understanding is that this will as well include inner scope of the outer function which is being completed, as the list is aggregated in a depth-first post-order:

```js
function foo() {
  for(let a of A) {
    bar(x => a + x);
  }
  for(let b of B) {
    bar(y => b + y);
  }
}
```

My point is that the scope of which defines `let a` which is supposed to be dumped when closing foo in post-order, would be closed when leaving the inner function `y => b + y`, as it would already be added to this list.

While I do not know if it matters after all, it does not seems to match the expectation of the comments which seems to suggest that only the inner scopes would be saved this way, not outer scopes.

Reply:
Here's my understanding of what should happen

```
function foo() {
  for(let a of A) {       <=  "a" is defined here
    bar(x => a + x);      <=  "a" is closed over by function here
  }                       <=  when leaving lexical scope for "for" here
                              "a" is added to the list, because it's
                              defined in this scope, and closed over by
                              inner script
  for(let b of B) {       <=  "b" is defined here
    bar(y => b + y);      <=  "b" is closed over by function here
  }                       <=  when leaving lexical scope for "for" here
                              "b" is added to the list, because it's
                              defined in this scope, and closed over by
                              inner script
}
```

So, the list should contain

["a", null, "b", null]

and gcthings list should contain

[function1. function2, "a", null, "b"]

where function1 is the index of `x => a + x`, and function2 is the index of `y => b + y`,
and trailing null is optimized out